### ISSUE TITLE
Multiclass Recognition issue when training with one class per image

### ISSUE THREAD

**Issue by EricK-Terra** [ultralytics/yolov5#4662]
## ❔Question

I am using yolov5 to realize recognition of 7 classes.

On my training/validation dataset each image has several objects (5 to 20) but only of one class.
After 200 epochs the results are good on the validation dataset

But when I run the model on "real" images with several classes the results are very bad.
Most of the time there is some kind of "area detection". that means that when a class is detected most of the objects around have the same class. So on an image there are most of the time only one or two classes detected and the objects dtected are grouped.

**Is there known issue when training with images with only one class objects ?**

## Additional context

I am using standard parameters for training and detection.
I am using yolov5s model with yolov5s.pt pre trained weights
I have 448 images for training and 74 images for classfication. those images contains around 10000 objects and the classes are not balanced (1 to 6 ratio)

My training/validation image size are 1920x1920 and I run training with an image size of 608
My "real" image size are 3120x3120 and contains around 50 objects

To summarize the differences between my training/validation images and "real" images :
### Training/validation images :
* one class per image
* between 10 and 20 objects
* space between objects
### "real images" :
* several classes per images
* around 50 objects per images
* object are close one to other and can in some case overlap

### metrics
At the end of 200 training epochs I have the following metrics :
* precision : 0.9
* recall 0.9
https://wandb.ai/erick-terra/YOLOv5/reports/Waste-Recognition--Vmlldzo5OTAxMDQ?accessToken=9fb61o4k16nz0ps4uejpbsw4ibvunk72z2jnulzcw5be1ocu773llywval5sg01s

**Comment by glenn-jocher** [comment]
@EricK-Terra your validation space should correlate with your deployment space if you expect your validation results to correlate with deployed results.

<img width="2029" alt="dataset" src="https://user-images.githubusercontent.com/26833433/132130143-5d694d09-053a-42cc-81c5-5ce823a8c2c7.png">

**Comment by lavrenkov-sketch** [comment]
@glenn-jocher May I have a little clarification. my images in val should be from the places where I want to recognize, however, the train should not contain images from the same environment, then I can train correctly and get a good Map?

**Comment by EricK-Terra** [comment]
Hello @glenn-jocher,
I understand the same as @lavrenkov-sketch from the schema.

So I have done a test with images from the "real world" in the validation dataset (that is to say images with mixed objects).
This time the mAP is low and both cls and obj loss function for validation dataset are increasing.
The results are visible here: https://wandb.ai/erick-terra/YOLOv5/runs/1pzqx6yc

So my question is : Is this bad result linked with the fact that in my training dataset if have only one class per image (so I shoul build a new training dataset) or is more linked to other issue (not enough images, .....) ?

**Comment by glenn-jocher** [comment]
@EricK-Terra 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/
